substrate/client/consensus/beefy/src/communication/fisherman.rs

```diff
@@ -0,0 +1,355 @@
+
+		let best_mmr_root = self
+			.runtime
+			.runtime_api()
+			.mmr_root(best_block_hash)
+			.map_err(|e| Error::RuntimeApi(e))?
+			.map_err(|e| Error::Backend(e.to_string()))?;
+
+		// if this errors, mmr has not been instantiated yet, hence the pallet is not active yet and
+		// we should not report equivocations
+		let leaf_count = self
+			.runtime
+			.runtime_api()
+			.mmr_leaf_count(best_block_hash)
+			.map_err(|e| Error::RuntimeApi(e))?
+			.map_err(|e| Error::Backend(e.to_string()))?;
+		let first_mmr_block_num = sp_mmr_primitives::utils::first_mmr_block_num::<B::Header>(
+			best_block_number,
+			leaf_count,
+		)
+		.map_err(|e| Error::Backend(e.to_string()))?;
+
+		if proof.commitment.validator_set_id != set_id ||
+			!proof.check::<BeefySignatureHasher, MmrHashing>(
+				best_mmr_root,
+				leaf_count,
+				&canonical_commitment_block_hash,
+				first_mmr_block_num,
+				best_block_number,
+			) {
+			debug!(target: LOG_TARGET, "🥩 Skip report for bad invalid fork proof {:?}", proof);
+			return Ok(false)
+		}
+
+		let offender_ids = proof.offender_ids();
+		if let Some(local_id) = self.key_store.authority_id(validator_set.validators()) {
+			if offender_ids.contains(&&local_id) {
+				warn!(target: LOG_TARGET, "🥩 Skip equivocation report for own equivocation");
+				return Ok(false)
+			}
+		}
+
+		let runtime_api = self.runtime.runtime_api();
+
+		let mut filtered_signatures = Vec::new();
+		// generate key ownership proof at that block
+		let key_owner_proofs: Vec<_> = offender_ids
+			.iter()
+			.cloned()
+			.filter_map(|id| {
+				match runtime_api.generate_key_ownership_proof(
+					canonical_commitment_block_hash,
+					set_id,
+					id.clone(),
+				) {
+					Ok(Some(proof)) => Some(Ok(proof)),
+					Ok(None) => {
+						debug!(
+							target: LOG_TARGET,
+							"🥩 Invalid fork vote offender not part of the authority set."
+						);
+						// if signatory is not part of the authority set, we ignore the signatory
+						filtered_signatures.push(id);
+						None
+					},
+					Err(e) => {
+						debug!(target: LOG_TARGET,
+							   "🥩 Failed to generate key ownership proof for {:?}: {:?}", id, e);
+						// if a key ownership proof couldn't be generated for signatory, we ignore
+						// the signatory
+						filtered_signatures.push(id);
+						None
+					},
+				}
+			})
+			.collect::<Result<_, _>>()?;
+
+		if key_owner_proofs.len() > 0 {
+			// filter out the signatories that a key ownership proof could not be generated for
+			let proof = ForkEquivocationProof {
+				signatures: proof
+					.signatures
+					.clone()
+					.into_iter()
+					.filter(|signature| !filtered_signatures.contains(&&signature.validator_id))
+					.collect(),
+				..proof
+			};
+			// submit invalid fork vote report at **best** block
+			runtime_api
+				.submit_report_fork_equivocation_unsigned_extrinsic(
+					best_block_hash,
+					proof,
+					key_owner_proofs,
+				)
+				.map_err(Error::RuntimeApi)?;
```

Comment:
From what I understand, the ancestry proofs, help us prove equivocation for blocks that are older than `current_block_number - 4096`. But will we be able to generate  `key_owner_proofs` for the validators associated with those blocks ?

Reply:
Yes, but proving equivocations of a prior session (or strictly speaking, for a block outside the node's pruning horizon) requires the fisherman to run an archival node - this is why the `canonical_commitment_block_hash` is passed into the applicable [`generate_ownership_proof` call](https://github.com/lederstrumpf/polkadot-sdk/blob/06492da09d0121d701182715a57a4b64b34fd65f/substrate/client/consensus/beefy/src/communication/fisherman.rs#L169-L173).

To remove the archival node requirement here, one could use an offchain worker - see the applicable note on [`BeefyApi::generate_key_ownership_proof`](https://github.com/lederstrumpf/polkadot-sdk/blob/06492da09d0121d701182715a57a4b64b34fd65f/substrate/primitives/consensus/beefy/src/lib.rs#L712-L717) (lifted from [`GrandpaApi::generate_key_ownership_proof`](https://github.com/lederstrumpf/polkadot-sdk/blob/06492da09d0121d701182715a57a4b64b34fd65f/substrate/primitives/consensus/grandpa/src/lib.rs#L523-L528)). See also the `Important` section of @andresilva's originating PR: https://github.com/paritytech/substrate/pull/3868.

Reply:
Oh, ok, that makes sense ! Sorry, I didn't know about the archival node requirement.